cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/resource/AbstractOperations.java

```diff
@@ -165,4 +167,68 @@ public T get(String namespace, String name) {
         return operation().inNamespace(namespace).withLabels(labels).list().getItems();
     }
 
+    /**
+     * Waits until resource is in the Ready state
+     *
+     * @param namespace The namespace
+     * @param name The resource name
+     * @param timeout Timeout in {@code timeUnit}
+     * @param timeUnit Time unit
+     */
+    public Future<Void> waitUntilReady(String namespace, String name, long timeout, TimeUnit timeUnit) {
+        Future<Void> fut = Future.future();
+        log.info("Waiting for {} resource {} in namespace {} to get ready", resourceKind, name, namespace);
+        AtomicInteger calls = new AtomicInteger(0);
+        long timer = 1000L;
+        long timeoutInMs = timeUnit.toMillis(timeout);
+
+        vertx.setPeriodic(timer, timerId -> {
+            calls.addAndGet(1);
+
+            vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
+                    future -> {
+                        try {
+                            R resourceOp = operation().inNamespace(namespace).withName(name);
+                            T resource = resourceOp.get();
+                            if (resource != null)   {
+                                if (Readiness.isReadinessApplicable(resource)) {
+                                    Boolean ready = resourceOp.isReady();
+
+                                    if (ready == true) {
+                                        future.complete();
+                                    } else {
+                                        future.fail("Not ready yet");
+                                    }
+                                } else {
+                                    future.complete();
+                                }
+                            } else {
+                                log.warn("{} {} in namespace {} doesn't exist", resourceKind, name, namespace);
```

Comment:
An alternative to failing if the resource doesn't exist is just to try again later and eventually fail with a timeout. If you did it that way then rather than waiting on the statefulset and then all the pods and then all the endpoints you could just bundle them all into one list and wait in the result. I'm not saying that's better necessarily, just pointing out an alternative.

Reply:
I kept that intentionally as a serial execution and not parallel. I was afraid that if one operation gets stuck somewhere, it would basically cut the timeout for the others. 

E.g. if the stateful set takes 50 seconds to get ready it would leave only 10 seconds to the Pods. This doesn't happen when I start `waitUntilReady` for Pods only once the StatefulSet is ready.